### ISSUE TITLE
Reporting accurate line numbers

### ISSUE THREAD

**Issue by davidtheclark** [stylelint/stylelint#133]
Progress:

* [x] `at-rule-empty-line-before`
* [x] `at-rule-no-vendor-prefix`
* [x] `block-closing-brace-newline-after`
* [x] `block-closing-brace-newline-before`
* [x] `block-closing-brace-space-after`
* [x] `block-closing-brace-space-before`
* [x] `block-no-empty`
* [x] `block-opening-brace-newline-after`
* [x] `block-opening-brace-newline-before`
* [x] `block-opening-brace-space-after`
* [x] `block-opening-brace-space-before`
* [x] `color-hex-case`
* [x] `color-hex-length`
* [x] `color-no-invalid-hex`
* [x] `comment-empty-line-before`
* [x] `comment-space-inside`
* [x] `custom-media-pattern`
* [x] `custom-property-no-outside-root`
* [x] `custom-property-pattern`
* [x] `declaration-bang-space-after`
* [x] `declaration-bang-space-before`
* [x] `declaration-block-semicolon-newline-after`
* [x] `declaration-block-semicolon-newline-before`
* [x] `declaration-block-semicolon-space-after`
* [x] `declaration-block-semicolon-space-before`
* [x] `declaration-colon-space-after`
* [x] `declaration-colon-space-before`
* [x] `declaration-no-important`
* [x] `function-calc-no-unspaced-operator`
* [x] `function-comma-space-after`
* [x] `function-comma-space-before`
* [x] `function-parentheses-space-inside`
* [x] `function-space-after`
* [x] `function-url-quotes`
* [x] `indentation`
* [x] `media-feature-colon-space-after`
* [x] `media-feature-colon-space-before`
* [x] `media-feature-name-no-vendor-prefix`
* [x] `media-feature-range-operator-space-after`
* [x] `media-feature-range-operator-space-before`
* [x] `media-query-list-comma-newline-after`
* [x] `media-query-list-comma-newline-before`
* [x] `media-query-list-comma-space-after`
* [x] `media-query-list-comma-space-before`
* [x] `media-query-parentheses-space-inside`
* [x] `nesting-block-opening-brace-newline-before`
* [x] `nesting-block-opening-brace-space-before`
* [x] `no-eol-whitespace`
* [x] `no-missing-eof-newline`
* [x] `no-multiple-empty-lines`
* [x] `number-leading-zero`
* [x] `number-max-precision`
* [x] `number-no-trailing-zeros`
* [x] `number-zero-length-no-unit`
* [x] `property-blacklist` - has tests, but no index required. Revert?
* [x] `property-no-vendor-prefix` - has tests, but no index required. Revert?
* [x] `property-whitelist` - has tests, but no index required. Revert?
* [x] `root-no-standard-properties` - has tests, but no index required. Revert?
* [x] `rule-nested-empty-line-before`
* [x] `rule-no-duplicate-properties`
* [x] `rule-no-shorthand-property-overrides`
* [x] `rule-no-single-line`
* [x] `rule-non-nested-empty-line-before`
* [x] `rule-properties-order`
* [x] `rule-trailing-semicolon`
* [x] `selector-combinator-space-after`
* [x] `selector-combinator-space-before`
* [x] `selector-list-comma-newline-after`
* [x] `selector-list-comma-newline-before`
* [x] `selector-list-comma-space-after`
* [x] `selector-list-comma-space-before`
* [x] `selector-no-attribute`
* [x] `selector-no-combinator` - https://github.com/postcss/postcss-selector-parser/issues/18
* [x] `selector-no-id` - https://github.com/postcss/postcss-selector-parser/issues/18
* [x] `selector-no-type` - https://github.com/postcss/postcss-selector-parser/issues/18
* [x] `selector-no-universal` - https://github.com/postcss/postcss-selector-parser/issues/18
* [x] `selector-no-vendor-prefix`
* [x] `selector-pseudo-element-colon-notation`
* [x] `selector-root-no-composition`
* [x] `string-quotes`
* [x] `value-list-comma-newline-after`
* [x] `value-list-comma-newline-before`
* [x] `value-list-comma-space-after`
* [x] `value-list-comma-space-before`
* [x] `value-no-vendor-prefix`

Using PostCSS's `result.warn()` we are associating warnings with nodes, and that means the message will have an *approximate* line number. That line number is only *approximate*, though, because nodes can contain multiple lines. Unless I'm missing something, then, relying on this as our mechanism for reporting the line numbers of errors will be inadequate.

It should not be a huge amount of trouble to calculate a more accurate line number, though: just take the starting line number of the node with the error and add however many newlines the start and the error.

What that does mean, though, unless I'm mistaken, is that we need to add extra information, beyond that provided by `result.warn()`. If we build our own warn function that *uses* `result.warn()`, we could do this. But then we run into a logging problem: if we allow PostCSS to report these linting errors as `warnings()`, they may frequently contain line number inaccuracies, right?

How do we address this?

Would it work if we could pass `.warn()` an *index within the node*, providing more specificity for the error than just the node itself?

For example, let's say I'm not allowing ids in selectors. Given
```css
.foo,
.bar,
#baz { ... }
```

I would want to say that my error is not just attached to the `rule` node, but it is more specifically attached to character ~15 of that node.

Or what if we passed `.warn()` line numbers *instead of* nodes?

One big downside to all of this is that we have to make sure that each error has an associated character, not just an associated node. In one way or another, we need that specific character in order to calculate a specific line, right? Which means that some of the rules that are super simple because they just use PostCSS node properties (e.g. `declaration-no-important`) would have to become a little more more complicated ...

**Comment by MoOx** [comment]
Maybe we can add raw message directly to the array ?
We also might ask for a change in PostCSS core for this (poke @ai)

**Comment by ai** [comment]
Yeap, wet can make `error()` and `warn()` smarter. What API do you want?

**Comment by davidtheclark** [comment]
My first thought is that we could add an `index` option to to `warn()`. `index` would be relative to the start of the node. You're already calculating a position for the start of each node; so if we then pass a relative `index`, it seems the position could be more accurately calculated accordingly.

E.g. for `declaration-no-important` applied to `a { color: pink !important; }
```js
result.warn("No important please", { node: decl, index: 12 });
```

When you access that warning to log it, then, it would have calculating the line/column for that index.

It seems to me that this should probably work in every case. Anybody have other ideas?

**Comment by MoOx** [comment]
We may be able to directly compute line/column with raw source (& so pass that as an argument for the warning), but index looks a good idea (+ simple to retrieve).

**Comment by davidtheclark** [comment]
Somebody is going to have to do that computation based on the raw source, yeah. I thought it might be common enough that PostCSS should do it instead of a plugin like stylelint doing it independently. *But* it might also be a useful API addition to allow the plugin to pass in a line/column position *in case* for some reason the plugin wants to do the calculation itself, in its own way -- right?

**Comment by MoOx** [comment]
Yep. PostCSS should probably allow anything to be overwritten from the 2nd arg.

**Comment by ai** [comment]
@davidtheclark we need `valuePos`, because `decl.between` can contains `\n` too.

**Comment by ai** [comment]
Please, create a issue in PostCSS repo. I agree that it will be useful (but I think you can miss it in first version of `stylelint` for faster release).

**Comment by lydell** [comment]
The multiple-newline-separated-selectors example could also benefit from postcss/postcss#76.

**Comment by MoOx** [comment]
I guess we can (for now) add a clear notice in the README so people won't come to complain about that.

**Comment by jeddy3** [comment]
Already done in #261 :)

**Comment by benfrain** [comment]
Unsure if the problem I have encountered is part of this or not? 

Just setting stylelint up for the first time and noticing that I'm not seeing the originating file in the output. Is it possible to do that? For example, I'm using postcss-import to concatenate my css files together and it would be ideal to have warnings provide the originating file-name too? Is this possible?

**Comment by davidtheclark** [comment]
@benfrain Are you using `postcss-reporter`, or are you printing the warnings by some other means?

**Comment by davidtheclark** [comment]
Also, maybe @MoOx knows more about this, could you use `postcss-import` *after* stylelint, instead of before, if it's `postcss-import` that is altering the reported source?

**Comment by benfrain** [comment]
@davidtheclark yes, using `postcss-reporter`. Config (via Gulp) looks like this:

````
gulp.task("styles", function () {
	var processors = [
		postcssImport({glob: true}),
		mixins,
		simplevars,
		nestedcss,
		postcssColorFunction(),
		mqpacker,
		stylelint({
			"rules": {
			   "rule-no-single-line": 0, // turn rule off
				"declaration-no-important": 2, // turn rule on
			}
		}),
		reporter({
			clearMessages: true,
		}),
		autoprefixer({browsers: ["last 2 version", "ios 6", "android 4"]}),
		csswring
	];
	return gulp.src("preCSS/styles.css")
	.pipe(postcss(processors).on("error", gutil.log))
	.pipe(gulp.dest("css/"))
	.pipe(reload({stream: true}));
});
````

Can't really move csslint before as import is the first thing that needs to happen. As you would imagine, if I do stick it before I just get nothing reported from stylelint.

**Comment by davidtheclark** [comment]
Oops, of course you're right you couldn't have done that. I was thinking wrong. I'll wait and see what @MoOx has to say about the source properties affected by `postcss-import`.

In the meanwhile, though, I think one temporary measure you could take would be to have a separate gulp task that runs the linting and make that run before `"styles"`. Something like

```js
var allCssFiles = ..;
var stylelintConfig = { .. };

gulp.task("lint-styles", function() {
  return gulp.src(allCssFiles)
    .pipe(postcss([
      stylelint(stylelintConfig),
      postcssReporter
    ]));
});

gulp.task("styles", ["lint-styles"], function() { .. });
```

Does that work for you?

**Comment by davidtheclark** [comment]
I'm actually doing that in my codebase currently: one linting step that grabs all the CSS files and runs them through both `stylelint` and `postcss-bem-linter`, and another processing step that does the other things. It's all fast enough for me that I don't care if the parser has to run twice.

**Comment by benfrain** [comment]
Sure, that could/should work, I'll give that a whirl. Thanks.

**Comment by benfrain** [comment]
@davidtheclark the only problem with this for me is that I use nesting so when running on my authoring style sheets stylelint tells me I'm using duplicate properties (for example I have font size set once, and then again inside a nested media query).
The more I think about it, the more I think that maybe there is no way it can run on the authoring styles for my scenario as I'm writing something that isn't actually CSS!

**Comment by davidtheclark** [comment]
Yeah, that would be a problem ... I'll open another issue and solicit ideas from postcss creators to see if there is (or can be in the future) a good way to handle this.

**Comment by MoOx** [comment]
For now postcss-import only make a low level replacement https://github.com/postcss/postcss-import/blob/e9db26da7afe8843af79df691c344b3d53b96e93/index.js#L453-L454
So it doesn't affect `.source` properties at all. So should not be responsible of anything.

**Comment by benfrain** [comment]
Does stylelint handle source maps? I'm just wondering if that's what I should use in this instance? That was it can run on the resultant CSS file but still provide details or the originating style sheet?

**Comment by MoOx** [comment]
stylelint is (for now) just a postcss plugin so I run on an AST. For now, it doesn't care about raw (original) css...

**Comment by davidtheclark** [comment]
@jeddy3: I checked off all the ones that I thought did not need indexes because the error could be attached to the start of the node. As for the ones I already specified locations for in tests --- who cares, may as well leave it. I also add a couple more commits to cover rules that I must have missed the first time around. Thanks for making the checklist!

**Comment by jeddy3** [comment]
> I checked off all the ones that I thought did not need indexes because the error could be attached to the start of the node.

Excellent :) Nice to see only a handful are blocked by the `postcss` and the `postcss-selector-parser` issues.

> As for the ones I already specified locations for in tests --- who cares, may as well leave it.

Sgtm :)

> I also add a couple more commits to cover rules that I must have missed the first time around

I saw those. They look good!

> Thanks for making the checklist!

No problemo :)
